pkg/kubelet/logs/container_log_manager.go

```diff
@@ -213,80 +223,121 @@ func (c *containerLogManager) Clean(ctx context.Context, containerID string) err
+	return nil
+}
+
+func (c *containerLogManager) processContainer(ctx context.Context, worker int) (ok bool) {
+	// Only rotate logs for running containers. Non-running containers won't
+	// generate new output, it doesn't make sense to keep an empty latest log.
+	// Note that we should not block log rotate for an error of a single container.
+	// In rotateLatestLog, there are several cases that we may
+	// lose original container log after ReopenContainerLog fails.
+	// We try to recover it by reopening container log.
+	// The container log should be recovered.
+	// Perform log rotation.
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer func() {
+		c.queue.Done(key)
+		c.queue.Forget(key)
+	}()
+	// Always default the return to true to keep the processing of Queue ongoing
+	ok = true
+	id := key.(string)
+
+	resp, err := c.runtimeService.ContainerStatus(ctx, id, false)
+	if err != nil {
+		klog.ErrorS(err, "Failed to get container status", "worker", worker, "containerID", id)
+		return
+	}
+	if resp.GetStatus() == nil {
+		klog.ErrorS(err, "Container status is nil", "worker", worker, "containerID", id)
+		return
+	}
+	path := resp.GetStatus().GetLogPath()
+	info, err := c.osInterface.Stat(path)
+
+	if err != nil {
+		if !os.IsNotExist(err) {
+			klog.ErrorS(err, "Failed to stat container log", "worker", worker, "containerID", id, "path", path)
+			return
 		}
-		if info.Size() < c.policy.MaxSize {
-			continue
+
+		if err := c.runtimeService.ReopenContainerLog(ctx, id); err != nil {
+			klog.ErrorS(err, "Container log doesn't exist, reopen container log failed", "worker", worker, "containerID", id, "path", path)
+			return
 		}
-		// Perform log rotation.
-		if err := c.rotateLog(ctx, id, path); err != nil {
-			klog.ErrorS(err, "Failed to rotate log for container", "path", path, "containerID", id)
-			continue
+
+		info, err = c.osInterface.Stat(path)
+		if err != nil {
+			klog.ErrorS(err, "Failed to stat container log after reopen", "worker", worker, "containerID", id, "path", path)
+			return
 		}
 	}
-	return nil
+	if info.Size() < c.policy.MaxSize {
+		klog.V(4).InfoS("log file doesn't need to be rotated", "worker", worker, "containerID", id, "path", path, "currentSize", info.Size(), "maxSize", c.policy.MaxSize)
+		return
+	}
+
+	if err := c.rotateLog(ctx, worker, id, path); err != nil {
+		klog.ErrorS(err, "Failed to rotate log for container", "worker", worker, "containerID", id, "path", path, "currentSize", info.Size(), "maxSize", c.policy.MaxSize)
+		return
+	}
+	return
 }
 
-func (c *containerLogManager) rotateLog(ctx context.Context, id, log string) error {
+func (c *containerLogManager) rotateLog(ctx context.Context, worker int, id, log string) error {
 	// pattern is used to match all rotated files.
 	pattern := fmt.Sprintf("%s.*", log)
 	logs, err := filepath.Glob(pattern)
 	if err != nil {
+		klog.ErrorS(err, "failed top list all log files with pattern", "worker", worker, "containerID", id, "path", log)
```

Comment:
Rather than a new log in all these places (presumably, these errors will eventually get logged), should we just move the additional data we're looking for in the `fmt.ErrorS(...)`? Not sure if worker number will be useful.

Reply:
Some of these logs were added only for me to help test this better. But I am open to suggestion on moving this to the error string itself. Or leave the logger here with some changes done to it as required.

Reply:
I'm not strongly opinionated here.